profiling/src/profile/mod.rs

```diff
@@ -430,14 +434,70 @@ impl Profile {
         // size of 32KiB should definitely out-perform starting at zero for
         // time consumed, allocator pressure, and allocator fragmentation.
         const INITIAL_PPROF_BUFFER_SIZE: usize = 32 * 1024;
-        let mut buffer: Vec<u8> = Vec::with_capacity(INITIAL_PPROF_BUFFER_SIZE);
-        profile.encode(&mut buffer)?;
+        let mut encoder = CompressedProtobufSerializer::with_capacity(INITIAL_PPROF_BUFFER_SIZE);
+
+        for (sample, timestamp, mut values) in std::mem::take(&mut self.observations).into_iter() {
+            let labels = self.translate_and_enrich_sample_labels(sample, timestamp)?;
+            let location_ids: Vec<_> = self
+                .get_stacktrace(sample.stacktrace)
+                .locations
+                .iter()
+                .map(Id::to_raw_id)
+                .collect();
+            self.upscaling_rules
+                .upscale_values(&mut values, &labels, &self.sample_types)?;
+
+            let item = pprof::Sample {
+                location_ids,
+                values,
+                labels,
+            };
+
+            encoder.encode(ProfileSamplesEntry::from(item))?;
+        }
+
+        // `Sample`s must be emitted before `SampleTypes` since we use
+        // the `sample_types` during upscaling.
+        // It is valid to emit protobuf fields out of order. See example in:
+        // https://protobuf.dev/programming-guides/encoding/#optional
```

Comment:
Minor: This comment makes sense, buttt I think there's an underlying assumption that's not clearly explained, which is that the iterators are consuming/releasing stuff as they are working, and thus if we were to reorder this to be the first step of the serialization process, we'd consume stuff right away.